rfcs/0016-error-handling.md

```diff
@@ -0,0 +1,379 @@
+
+Therefore, by extending `assert` with the `error` parameter, and promoting the usage of `assert_(n)eq`, we essentially enforce via API the best practice of providing helpful debugging information.
+
+As a part of the consolidation, the existing contextless `assert(<v1> == <v2>)` lines in tests will be replaced with troubleshooting friendly `assert_eq(<v1>, <v2>)` for all `AbiEncode`able types.
+
+The functions in `std::assert` will be rewritten to use `panic` internally. This will remove the need for hardcoded `ASSERT_SIGNAL`s constant, provide logged information tied to the error (currently, in case of additional logs, developers need to know to look at the last logs), and allow usage of the new `#[test]` arguments (`panic_msg` and `panic_value`). An error type enum `AssertError` will be introduced:
+
+```sway
+pub struct AssertEq<T> where T: AbiEncode {
+    pub expected: T,
+    pub actual: T,
+}
+
+#[error_type]
+pub enum AssertError<T> where T: AbiEncode {
+    #[error(m = "`std::assert::assert` has failed.")]
+    Assert: T,
+    #[error(m = "`std::assert::assert_eq` has failed.")]
+    AssertEq: AssertEq<T>,
+    #[error(m = "`std::assert::assert_neq` has failed.")]
+    AssertNeq: T,
+}
+```
+
+This means that a typical failing assert will result in an error message like:
+
+```
+`std::assert::assert` has failed.
+└ This is the exact user-defined helpful message.
+```
+
+Note that in this case we panic on error type enums and not on hardcoded string slices, thus introducing overhead. This is not in the collision with the second point in the guidelines (zero-cost), because:
+- end-developers can always choose to use `if + panic` directly instead of `assert`.
+- we assume `assert_eq` and `assert_neq` to be used exclusively in tests and not on-chain.
+
+### `require`
+
+Having `assert<T>(condition: bool, error: T) where T: Error` eliminates the need for a separate `require` function, or at least makes the existing question of the difference between `assert` and `require` more prominent.
+
+The proposal is to deprecate `require` and replace it by `assert<T>(condition: bool, error: T) where T: Error`. This proposal is controversial. The argumentation for it is the following. `assert` is Rust inspired API and is used both in tests and non-test code. `require` is intended to check pre- or post conditions and is inspired by Solidity's `require` function. It the context of blockchain development in Rust it is used as well, e.g., in the [anchor_lang](https://docs.rs/anchor-lang/latest/anchor_lang/index.html), Rust eDSL for writing Solana programs.
+
+Having both `assert` and `require` introduces confusion on intended usage. Should `assert` be used only in tests? Or also for catching internal errors in non-test code? If so, we have it currently inconsistently used even in the `std`, e.g., in the `std::math`.
```

Comment:
Perhaps a less agressive migration would be to have them alias each other, since they are used interchangeably?